sql/2020/02_JavaScript/avg_pct_per_page_scripts_using_async_defer_module_nomodule.sql

```diff
@@ -0,0 +1,44 @@
+#standardSQL
+# Average Percent per page of external scripts using Async, Defer, Module or NoModule attributes.
+SELECT 
+  client,
+  AVG(pct_external_async) AS avg_pct_external_async,
+  AVG(pct_external_defer) AS avg_pct_external_defer,
+  AVG(pct_external_module) AS avg_pct_external_module,
+  AVG(pct_external_nomodule) AS avg_pct_external_nomodule
+FROM
+(
+  SELECT 
+    client,
+    page,
+    COUNT(*) as external_scripts,
+    SUM(IF(script LIKE "%async%",1,0)) AS async,
+    SUM(IF(script LIKE "%defer%",1,0)) AS defer,
+    SUM(IF(script LIKE "%module%",1,0)) AS module,
+    SUM(IF(script LIKE "%nomodule%",1,0)) AS nomodule,
+    SUM(IF(script LIKE "%async%",1,0))    / COUNT(*) AS pct_external_async,
+    SUM(IF(script LIKE "%defer%",1,0))    / COUNT(*) AS pct_external_defer,
+    SUM(IF(script LIKE "%module%",1,0))   / COUNT(*) AS pct_external_module,
+    SUM(IF(script LIKE "%nomodule%",1,0)) / COUNT(*) AS pct_external_nomodule
+  FROM 
+  (
+    SELECT  
+      _TABLE_SUFFIX AS client,
+      page, 
+      url,
+      REGEXP_EXTRACT_ALL(LOWER(body), "(<script [^>]*)") AS scripts
+    FROM 
+      `httparchive.response_bodies.2020_08_01_*` 
+    WHERE  
+      LOWER(body) LIKE "%<html%"
```

Comment:
@paulcalvano FYI this is super expensive because it has to look at all response bodies. The `almanac.summary_response_bodies` table is clustered by `firstHtml` so we can restrict the query to only HTML files
```suggestion
    FROM 
      `httparchive.almanac.summary_response_bodies` 
    WHERE
      date = '2020-08-01' AND
      firstHtml
```

Reply:
@paulcalvano I've updated this query to use the `almanac` dataset approach and updated the results in the sheet. This saves about 24 TB per query.

I've also made a couple of other changes:
- removed some unused fields
- changed `LIKE` to `REGEXP_CONTAINS` so we can check attribute boundaries 
- nit fixes like uppercasing `as` keywords

Could you apply this approach to the other queries that depend on response body data and update the results?

Reply:
Thanks.  That reduced those queries from 25TB to 1.5TB !